Development of the home page

Danila-Peshev opened the pull request:
Implementation of the main page and all her components

Danila-Peshev committed:
Commit [0]:
Rename fields User;
Add watchedMoviesId field.

Danila-Peshev committed:
Commit [1]: Files structuring

Danila-Peshev committed:
Commit [3]:
Update UserRepository(like back);
Add UserService;

Danila-Peshev committed:
Commit [2]:
Move main indexes it as it was;

Danila-Peshev committed:
Commit [4]:
Update locales;
Create global NavBar;
Move LanguageSwitcher on the NavBar;

Danila-Peshev committed:
Commit [5]:
Logic for the fields of user has been moved to AuthContext;
Add unfinished blocks with genres and movies;
Add button images in assets;
Implemented block view;

Danila-Peshev committed:
Commit [6]:
Update ProtectedRoute(before the update, he made a request if the user is not logged in);
Update MovieService and fields of type Movie(due to the difference in the names of the type fields and the request response, some fields were undefined);
Сomponents of the home page are divided into blocks and internationalization has been added to them;
Added ViewContext, and PaginationBlock.

Danila-Peshev committed:
Commit [7]:
Fixed updating of the entire main block when selecting genres.

Isidora25 commented on src/assets/views-images/block_view.png:
для иконок лучше использовать svg 
они не теряют в качестве и всегда будут выглядеть хорошо
при этом не будут занимать много места

Isidora25 commented on src/assets/views-images/row_view.png:
то же самое про svg

Isidora25 commented on src/global-components/NavBar.tsx:
а чем global-components отличаются от components? 
обычно если components папка лежит сразу в src, это уже означает, что компоненты глобального уровня 
все неглобальные компоненты должны являться подпапками тех страниц/компонентов, к которым они относятся

Isidora25 commented on src/components/movies-block/MoviesBlock.tsx:
```diff
@@ -0,0 +1,130 @@
+  const [page, setPage] = useState<number>(1);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string | null>(null);
+  const { user, getWatchedMoviesId, getFavoriteMoviesId } = useAuth();
```
непонятно, почему контекст аутентификации, возвращает еще и айдишники фильмов 
это совершенно другой смысловой контекст

Isidora25 commented on src/components/movies-block/MoviesBlock.tsx:
```diff
@@ -0,0 +1,130 @@
+  const { language } = useLanguage();
+  const totalPages = Math.min(
+    moviesResponse ? moviesResponse.totalPages : 1,
+    500
```
почему 500? почему не 5? почему не 10 миллионов?

Isidora25 commented on src/components/movies-block/MoviesBlock.tsx:
```diff
@@ -0,0 +1,130 @@
+    setIsLoading(true);
+    setError(null);
+    try {
+      if (isOnAddedPage) {
```
мне очень непонятно, что такое AddedPage. Добавленная страница? куда она добавлена?
переименуй, пожалуйста

Danila-Peshev replied:
Имеется ввиду - блок на странице добавления фильмов, если да то убираются кнопки у фильмов. Если честно не знаю как правильно это заименовать, подскажи пожалуйста как лучше)

Isidora25 commented on src/components/movies-block/MoviesBlock.tsx:
```diff
@@ -0,0 +1,130 @@
+    setIsLoading(true);
+    setError(null);
+    try {
+      if (isOnAddedPage) {
+        const response = await getMoviesByFilters({
+          language,
+          activePage: page,
+        });
+        setMoviesResponse(response);
+      } else {
+        const favoriteMoviesIds = getFavoriteMoviesId();
+        if (favoriteMoviesIds.length === 0) {
+          setIsLoading(false);
+          setMoviesResponse({
+            page: 1,
+            results: [],
+            totalPages: 1,
+            totalResult: 0,
+          });
+          return;
+        }
```
слишком усложняешь компонент. он получается слишком многозадачным. у компонента/класса/функции в идеале (к которому стоит стремиться) должна быть одна задача. в твоем случае, задаче компонента - отобразить список фильмов. компоненту должно быть все равно, что это за фильмы - избранные, не избранные. другиими словами, тебе нужно вынести логику запроса фильмов на уровень выше

Isidora25 commented on src/components/movies-block/OneMovie.tsx:
```diff
@@ -0,0 +1,164 @@
+  isOnAddedPage: boolean;
+}
+
+const MovieBlock: FC<MovieBlockProps> = ({
```
файл называется OneMovie, а сам компонент иначе. Переименуй, плиз

Isidora25 commented on src/components/movies-block/OneMovie.tsx:
```diff
@@ -0,0 +1,164 @@
+
+  if (isBlockView) {
+    return (
+      <div className="flex flex-row gap-3">
+        <span className="text-gray-400 w-5">{movieNumber}.</span>
+        <div
+          key={simpleMovie.id}
+          className="relative rounded bg-gray-800 text-white p-4 w-80"
+        >
+          {isWatched ? (
+            <div className="absolute rounded inset-0 bg-gray-800 opacity-70 z-10"></div>
+          ) : (
+            <></>
+          )}
+          <img
+            className="rounded"
+            alt={simpleMovie.title}
+            src={`https://image.tmdb.org/t/p/w400${simpleMovie.posterPath}`}
+          />
+          <div className="w-11/12">
+            <h3 className="text-xl font-semibold">{simpleMovie.title}</h3>
+            <p className="text-sm text-gray-400">
+              {simpleMovie.releaseDate.split("-")[0]}
+            </p>
+            <p className="text-sm text-gray-300">{simpleMovie.overview}</p>
+            <div className="flex flex-wrap my-2">
+              {simpleMovie.genreIds.map((genreId) => {
+                const genre = genres.find((genre) => genre.id === genreId);
+                return (
+                  <span
+                    key={genreId}
+                    className="bg-gray-700 text-sm rounded-full px-2 py-1 mr-2 mb-2"
+                  >
+                    {genre ? genre.name : "Unknown Genre"}
+                  </span>
+                );
+              })}
```
Это очень сложно читать 
либо вынеси в компонент, который принимает проп isBlockView и меняет внешний вид, когда isBlockView меняется
либо это два разных компонента, один для блочного вида, второй для списка

Isidora25 commented on src/components/movies-block/OneMovie.tsx:
```diff
@@ -0,0 +1,164 @@
+            </div>
+          </div>
+        </div>
+        {isOnAddedPage ? (
+          <></>
+        ) : (
+          <div className="z-20 space-y-4">
+            <button
+              onClick={handleClickWatched}
+              className={`h-6 w-6 ${
+                isWatched
+                  ? "bg-green-800 text-white"
+                  : "bg-white text-green-900"
+              }`}
+            >
+              ✔
+            </button>
+            <button
+              onClick={() => removeFavoriteMovie(simpleMovie.id)}
+              className="h-6 w-6 bg-red-700 hover:bg-red-950"
+            >
+              ✖
+            </button>
+          </div>
+        )}
+      </div>
```
condition ? <div>...</div> : null

Isidora25 commented on src/components/movies-block/PaginationBlock.tsx:
```diff
@@ -0,0 +1,61 @@
+  setPage,
+}) => {
+  const { t } = useTranslation();
+  const createPageNumbers = () => {
+    const pages = [];
+    const startPage = Math.max(1, page - 5);
+    const endPage = Math.min(totalPages, page + 5);
+
+    for (let i = startPage; i <= endPage; i++) {
+      pages.push(i);
+    }
+
+    return pages;
+  };
```
это функция хелпер, вынеси ее из тела компонента, иначе она будет пересоздаваться при каждом ререндере

Isidora25 commented on src/components/movies-block/PaginationBlock.tsx:
```diff
@@ -0,0 +1,61 @@
+          onClick={() => setPage(1)}
+          className="text-white bg-blue-800 rounded-sm px-4 py-1"
+        >
+          {t("firstPage")}
```
зачем так сложно) 
арабские цифры прекрасно справляются с индикацией номера страницы)

Isidora25 commented on src/pages/home/GenresBlock.tsx:
```diff
@@ -0,0 +1,51 @@
+    <div className="w-full">
+      <span>{t("selectYouFavoriteGenres")}</span>
+      <div className="mt-5 grid grid-cols-10 grid-rows-2 gap-2">
+        {genres.map((genre) => (
+          <button
+            className={`hover:bg-blue-950 hover:text-white rounded-full px-5 py-2 overflow-hidden truncate ${
+              favoriteGenresId.includes(genre.id)
+                ? "bg-white text-blue-950"
+                : "bg-blue-700 text-white"
+            }`}
+            onClick={() => handleClickOnGenre(genre.id)}
+          >
+            {genre.name}
+          </button>
+        ))}
```
не забывай про атрибут key

Danila-Peshev committed:
correction:
Splitting a block with one movie into two;
Moved the logic of the user's fields to the UserContext;
Now MoviesBlock only accepts movies;
Pagination Black has been changed - the page counting function has been removed, now the first and last pages are displayed in numbers and only when they are not visible;
A constant of 500 pages has been introduced;
Icons are now in svg;
In the files, simpleMovie/s is replaced by movie/s.

Danila-Peshev committed:
correction:
Check for Movies Response has been removed, the default result is empty.

Isidora25 commented on src/components/UserContext.tsx:
```diff
@@ -40,35 +27,10 @@ const defaultAuthContext: AuthContextType = {
-    localStorage.removeItem("user");
-    setUser(null);
-  };
+export const UserProvider = ({ children }: { children: ReactNode }) => {
```
а зачем тебе этот контекст, если тут нет ни одного своего стейта? 
вынеси просто все эти функции в сервис - общий файл и импортируй их точно так же как этот контекст 
контекст нужен для хранения глобального стейта, здесь он выглядит излишним

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -1,47 +1,114 @@
+  const { getFavoriteMoviesId } = useUser();
   const { t } = useTranslation();
 
+  const totalPages = Math.min(
+    moviesResponse ? moviesResponse.totalPages : 1,
+    MAX_TOTAL_PAGES
+  );
```
а зачем оно?

Danila-Peshev replied:
Ответ от tmdb может вернуть больше 1 тыс. страниц, а когда обращаешься к ним со страницей больше 500 они возвращают invalid page error 22, поэтому сделал ограничение.

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -1,47 +1,114 @@
+        page,
+      });
+      setMoviesResponse(movies);
+    } catch (err) {
+      setStatus((prevStatus) => ({ ...prevStatus, errorMovies: true }));
+    } finally {
+      setStatus((prevStatus) => ({ ...prevStatus, loadingMovies: false }));
```
обработку ошибок от апи лучше спрятать в сервис работы с апи и возвращать эти ошибки сразу оттуда. если они тут нужны, конечно же

Danila-Peshev replied:
Хорошо, я их добавил, потому что на некоторых браузерах без VPN обрывало попытку запроса, и чтобы юзер не ждал вечной загрузки, решил обрабатывать ошибки

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -1,47 +1,114 @@
 
+  useEffect(() => {
+    fetchMoviesResponse();
+  }, [page, user?.favoriteMoviesId]);
```
не нужно хранить инфо по любимым жанрам, фильмам в объекте юзер 
да, они зависят от юзера, но это разные области

Danila-Peshev replied:
Мне получается нужно из localStorage сделать бд?
Просто для этого потребуется сделать три таблицы: юзеры, юзерЖанр и юзерФильм.
Т.к. это LS придется обновить функции получения жанров и фильмов по id юзера.
Мне сделать это, или лучше подождать пока дойду до бэка?

Isidora25 replied:
нет, тебе просто не нужно возвращать это в объект юзер, функция getFavMoviesIds должна возвращать только айдишники фильмов, без объекта юзера

Danila-Peshev replied:
Эти функции возвращают массив чисел, без объекта юзера.
А функции с добавлением/удалением обновляют данные в LS.
Если сделать всё без юзера, то в LS придётся создать связные таблицы, где в аргументах функциях будут крутиться только id юзеров.

Мне тяжело это понять, просто я привык, что по запросам указывается id пользователя и возвращается в ответе массив id жанров/фильмов и готово.
Не понимаю как это сделать без участия юзера, если нужно получить id избранных фильмов/жанров по id этого юзера.

Как нужно сделать, объясни подробнее пожалуйста, я потерялся в этом вопросе 😥

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -1,47 +1,114 @@
     <ViewProvider>
       <div className="w-11/12 flex flex-col gap-y-5 p-5 mx-auto mt-24 mb-12 text-white text-base font-medium bg-gray-900 rounded">
-        {loading ? (
+        {loadingGenres || loadingMovies ? (
```
зачем тебе два разных стейта для лоадинга, если ты их используешь одновременно?

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -1,47 +1,114 @@
+        {loadingGenres || loadingMovies ? (
           t("loading")
-        ) : error ? (
+        ) : errorGenres || errorMovies ? (
```
тот же вопрос про ошибку, но от стейта для ошибки лучш вообще избавиться

Isidora25 commented on src/pages/home/movies-block/BlockViewOneMovie.tsx:
```diff
@@ -0,0 +1,87 @@
+          </div>
+        </div>
+      </div>
+      {showAdditionalButtons ? (
+        null
+      ) : (
+        <div className="z-20 space-y-4">
+          <button
+            onClick={handleClickWatched}
+            className={`h-6 w-6 ${
+              isWatched
+                ? "bg-green-800 text-white"
+                : "bg-white text-green-900"
+            }`}
+          >
+            ✔
+          </button>
+          <button
+            onClick={() => removeFavoriteMovie(movie.id)}
+            className="h-6 w-6 bg-red-700 hover:bg-red-950"
+          >
+            ✖
+          </button>
+        </div>
+      )}
```
а почему наоборот? 
если show, то обычно показывают что-то, а у тебя null)

Isidora25 replied:
либо переименуй флаг в hide, либо поправь тернарник

Isidora25 commented on src/pages/home/movies-block/RowViewOneMovie.tsx:
```diff
@@ -0,0 +1,89 @@
+  };
+
+  return (
+    <div
```
очень много одинакового кода, как в BlockViewOnMevie 
точно нельзя стилями обыграть два разных вида? 
у тебя нет никаких ограничений в виде мокапов от дизайнеров, сделай себе жизнь легче и сделай эти два комопнента максимально одинаковыми, только меняй расположение контента 
если вдруг тебе понадобится добавить новый элемент для отоборажения фильма, тебе придется делать это дважды
если компонент будет один, то всего лишь один раз

Danila-Peshev replied:
Могу, но только в className контейнеров везде будут проверки на isBlockView, это правильный подход, или есть другой способ?

Isidora25 replied:
придумай себе дизайн такой, чтобы от isBlockView зависели только стили корневого контейнера

Danila-Peshev committed:
commit:
Changed the AuthContext so that everyone does not check for the presence of a user every time, ProtectedRoute and NavBar are changed according to it;
The Rop and Block blocks have been combined and the design has been changed;
Tables have been created in LS to store user data;
Instead of the user's context, a service has been created that interacts with LS through the user's ID;
The processing of all button actions has been moved to the main Home component;
Everything works as it should, but when you click on the "viewed" and "remove" buttons, the entire page is re-rendered and the user is moved up the page.

Isidora25 commented on src/components/AuthContext.tsx:
```diff
@@ -9,54 +9,60 @@ import { User } from "../types/User";
 
 export const AuthProvider = ({ children }: { children: ReactNode }) => {
-  const [user, setUser] = useState<User | null>(null);
+  const [user, setUser] = useState<User>({} as User);
```
а зачем тебе тут пустой объект? нормально же все было с null
а сейчас всякие костыли лезут в виде проверки наличия ключей объекта

Isidora25 commented on src/components/AuthContext.tsx:
```diff
@@ -9,54 +9,60 @@ import { User } from "../types/User";
 };
 
-export const useAuth = () => useContext<AuthContextType>(AuthContext);
+export const useAuth = () => useContext(AuthContext);
```
почему убрал дженерик?

Danila-Peshev replied:
Не увидел в попытках менять AuthContext, вернул.

Isidora25 commented on src/services/UserDataService.ts:
```diff
@@ -0,0 +1,113 @@
+  const userGenres = JSON.parse(localStorage.getItem("userGenres") || "[]");
+  return userGenres
+  .filter((value: { user_id: number; }) => value.user_id === userId)
+  .map((value: { genre_id: any; }) => value.genre_id);
```
избавься от any плиз

Danila-Peshev committed:
commit:
Return User or null.

reshetov commented on src/pages/home/movies-block/OneMovie.tsx:
```diff
@@ -0,0 +1,94 @@
+        <img
+          className={`rounded ${isBlockView ? null : `relative h-40`}`}
+          alt={movie.title}
+          src={`https://image.tmdb.org/t/p/w400${movie.posterPath}`}
```
Передвинь в константу или вообще в env

Danila-Peshev committed:
Took out the url path to the image in the env file.

Danila-Peshev committed:
del add page
